### PR TITLE
Enhance bpf interface autodetection

### PR DESCRIPTION
**Pull request by sridhartigera**
## Description

This PR enhances the host interface auto-detection logic for the eBPF dataplane.
with the new approach, BPF endpoint manager builds a set of trees based on the netlink information of each host interface.

For example, if we have 2 interfaces in a bond and vlan on top of the bond, vlan interface is at the root of the tree, bond is at level 1 and physical interfaces are at the leaves. In the case of single standalone interfaces, each interface is a single node tree. Attaching Tc,XDP programs follows the below logic
1. If the interface is a root node, attach Tc.
2. If the interface is a leaf node, attach XDP.
3. If the interface is neither root nor leaf, it is ignored.

In the above process, XDP and Tc programs are removed from the interfaces when the interface moves down the tree.

```
bpftool net
xdp:
enp1s0f0np0(2) driver id 965
enp1s0f1np1(4) driver id 968

tc:
lo(1) clsact/ingress cali_tc_preambl:[804] id 804
lo(1) clsact/egress cali_tc_preambl:[805] id 805
bond0.1001(6) clsact/ingress cali_tc_preambl:[974] id 974
bond0.1001(6) clsact/egress cali_tc_preambl:[976] id 976
bpfout.cali(7) clsact/ingress cali_tc_preambl:[844] id 844
bpfout.cali(7) clsact/egress cali_tc_preambl:[843] id 843
cali662a7526402(9) clsact/ingress cali_tc_preambl:[894] id 894
cali662a7526402(9) clsact/egress cali_tc_preambl:[895] id 895
vxlan.calico(10) clsact/ingress cali_tc_preambl:[823] id 823
vxlan.calico(10) clsact/egress cali_tc_preambl:[822] id 822
cali596936e74c1(11) clsact/ingress cali_tc_preambl:[896] id 896
cali596936e74c1(11) clsact/egress cali_tc_preambl:[889] id 889
calie500d5708cd(12) clsact/ingress cali_tc_preambl:[890] id 890
calie500d5708cd(12) clsact/egress cali_tc_preambl:[891] id 891
```

xdp attached to physical interface and tc to bond.vlan

## Todos

- [x] Tests
- [ ] Documentation
- [x] Release note

## Release Note

```release-note
ebpf: improved autodetection of device hierarchy like bonds and vlans. tc is attached to the root and xdp to the leaves.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 1]
```diff
@@ -1122,87 +1131,37 @@ func (m *bpfEndpointManager) onInterfaceUpdate(update *ifaceStateUpdate) {
 		return
 	}
 
+	// Should be safe without the lock since there shouldn't be any active background threads
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:1134]: why did you move this down here?

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 2]
```diff
@@ -1122,87 +1131,37 @@ func (m *bpfEndpointManager) onInterfaceUpdate(update *ifaceStateUpdate) {
...
+			// update the ifaceType, master ifindex if bond slave.
+			link, err := m.dp.getIfaceLink(update.Name)
 			if err != nil {
-				log.WithError(err).Warn("Failed to get master interface name. Slave devices not updated")
-			} else {
-				// Slave interface update comes first.
-				if _, ok := m.hostIfaceToSlaveDevices[netiface.Name]; !ok {
-					m.hostIfaceToSlaveDevices[netiface.Name] = set.New[string]()
+				log.Errorf("Failed to get interface information via netlink '%s'", update.Name)
+				curIfaceType = IfaceTypeL3
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:1148]: why do we default to L3?

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 3]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+			eintf.name = intf.name
+			eintf.index = intf.index
+			eintf.masterIndex = intf.masterIndex
+			eintf.parentIndex = intf.parentIndex
+		}
+		if val, exists := m.hostIfaces[attrs.Index]; !exists {
+			m.hostIfaces[attrs.Index] = intf
+		} else {
+			val.name = intf.name
+			val.index = intf.index
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4255]: it must already have the same index ;-)

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 4]
```diff
@@ -212,6 +212,15 @@ var zeroIface bpfInterface = func() bpfInterface {
 	return i
 }()
 
+type bpfHostIface struct {
+	name        string
+	index       int
+	masterIndex int
+	parentIndex int
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:219]: can the interface have both master and parent in the tree? It feels to me that that the system ifaces can have either or, but not both and it is just a naming convention. I wonder if it would make thing simple if we maintained only a parent for both in our internal tree. Or maybe I am confused since there is just one parent pointer.

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 5]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+			m.hostIfaces[attrs.Index] = intf
+		} else {
+			val.name = intf.name
+			val.index = intf.index
+			val.masterIndex = intf.masterIndex
+			val.parentIndex = intf.parentIndex
+		}
+	} else if attrs.MasterIndex != 0 {
+		m.deleteHostIface(intf.name)
+		// If the master interface is in the tree, add it as a slave.
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4261]:
```suggestion
		// If the master interface is a root, add it as a slave.
```

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 6]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+		if val, exists := m.hostIfaces[attrs.MasterIndex]; exists {
+			intf.parentIface = val
+			val.children[attrs.Index] = intf
+			m.hostIfaces[attrs.MasterIndex] = val
+		} else {
+			// Need to search slave devices.
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4267]:
```suggestion
			// Need to search the tree to find the master
```

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 7]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+				intf.parentIface = masterIf
+				masterIf.children[attrs.Index] = intf
+			} else {
+				masterIface := &bpfHostIface{index: attrs.MasterIndex, children: make(map[int]*bpfHostIface)}
+				intf.parentIface = masterIface
+				masterIface.children[attrs.Index] = intf
+				m.hostIfaces[attrs.MasterIndex] = masterIface
+			}
+		}
+		delete(m.hostIfaces, attrs.Index)
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4279]: why are you deleting it if you just added it and it could be the new root?

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 8]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+	} else if attrs.ParentIndex != 0 {
+		// bond0 already in list
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4281]: bond0 comment is irrelevant

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 9]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+	}
+}
+
+func isLeafIface(hostIf *bpfHostIface) bool {
+	for _, child := range hostIf.children {
+		if child.name != "" {
+			return false
+		}
+	}
+	return true
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4435]:
```suggestion
	return len(hostIf.children) == 0
```
perhaps?

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 10]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+	for _, iface := range m.hostIfaces {
+		slaveIf, _ := findInIfaceTree(iface, -1, name, false)
+		if slaveIf != nil {
+			return slaveIf
+		}
+	}
+	return nil
+}
+
+func getAllIfaces(root *bpfHostIface, nodes *[]string) {
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4419]:
```suggestion
func getAllIfaces(root *bpfHostIface) []string {
```
perhaps?

**Review by tomastigera — CHANGES_REQUESTED** [review]

File: felix/dataplane/linux/bpf_ep_mgr.go (hunk 9 above)
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:4447]: In our FVs, eth0 is a veth which has a parent index. Hence the parent interface will be added as a child but it wont have a name, since felix wont get the update. This results in len(hostIf.children) > 0 but its a invalid child. Hence I did not do this way.

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 11]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
 }
 
 func (m *bpfEndpointManager) getBondSlaves(masterIfName string) set.Set[string] {
-	if val, ok := m.hostIfaceToSlaveDevices[masterIfName]; ok {
+	slaveDevices := set.New[string]()
+	link, err := m.dp.getIfaceLink(masterIfName)
+	if err != nil {
+		log.Errorf("Error getting link for interface %s, err = %v", masterIfName, err)
+		return slaveDevices
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go]: Have changed to return an empty slice of string.

File: felix/dataplane/linux/bpf_ep_mgr.go (hunk 7 above)
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go]: When adding an interface with masterIndex !=0 (which means the intf is a slave intf), we need to delete the interface from the tree before adding it because the interface might have moved from standalone to a bond. Hence the deletion.

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 12]
```diff
@@ -305,6 +314,7 @@ type bpfEndpointManager struct {
 	ifStateMap              *cachingmap.CachingMap[ifstate.Key, ifstate.Value]
 	removeOldJumps          bool
 	legacyCleanUp           bool
+	hostIfaces              map[int]*bpfHostIface
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go]: changed it to hostIfaceTrees and introduced a new type bpfIfaceTrees

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 13]
```diff
@@ -4201,10 +4211,246 @@ func (m *bpfEndpointManager) getIfaceTypeFromLink(link netlink.Link) IfaceType {
...
+	attrs := link.Attrs()
+	intf := &bpfHostIface{name: attrs.Name,
+		index:       attrs.Index,
+		masterIndex: attrs.MasterIndex,
+		parentIndex: attrs.ParentIndex,
+		children:    make(map[int]*bpfHostIface)}
+
+	if attrs.MasterIndex == 0 && attrs.ParentIndex == 0 {
+		eintf := m.findHostIfaceByIndex(attrs.Index)
+		if eintf != nil {
```
[the remark this answers is not in the document]
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go]: I just rewrite the name so that I don't overwrite the children information. For example, when felix gets the child information first, we create the child node and also the parent node with master index. Now when felix gets the update of the parent interface, the interface with that index is already there in the tree. Just need to update the name.

File: felix/dataplane/linux/bpf_ep_mgr.go (hunk 2 above)
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:1149]: when we don't get the link information, I default the type to L3 and if the interface matches in `dataIfacePattern`, I change it to Data.

File: felix/dataplane/linux/bpf_ep_mgr.go (hunk 1 above)
Comment (sridhartigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:1135]: I take the lock in `cleanupOldAttach` to get the policyIdx info from the state. Hence moved the lock in the calling function below the block where we call `cleanupOldAttach`

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 14]
```diff
@@ -4371,62 +4420,25 @@ func (m *bpfEndpointManager) getRootInterface(hostIf *bpfHostIface) *bpfHostIfac
...
-	}
-}
-
-func isLeafIface(hostIf *bpfHostIface) bool {
+// isLeafIface returns if the interface does not have a valid child.
+// This can be as simple as return len(hostIf.children) == 0.
+// In the case of FVs, eth0 is a veth which has a parent interface.
+// Though eth0 is the only valid interface in the tree, we add the
+// parent interface as a child, Hence we need to be sure if there are
+// valid child interfaces.
```
Comment (tomastigera) [inline felix/dataplane/linux/bpf_ep_mgr.go:4440]:
```suggestion
// However, the main interfaces may have a parent in a different namespace, like a veth,
// and we are not going to see and update from that parent.
```

**Review by tomastigera — CHANGES_REQUESTED** [review]
I have come mostly naming nits, lgtm in general

**Review by tomastigera — APPROVED** [review]
